File: Apps/SGM/sportsbet_sgm.py
import requests
import json
import pandas as pd
import os
from pymongo import MongoClient

#==============================================================================
# API Call function
#==============================================================================

# Create function to call the API
def get_sgm_sportsbet(data, player_names, number_of_disposals):
    if len(player_names) != len(number_of_disposals):
        raise ValueError('Both lists should have the same length')

    filtered_df = pd.DataFrame()
    for i in range(len(player_names)):
        temp_df = data[(data['player_name'] == player_names[i]) & 
                                  (data['number_of_disposals'] == number_of_disposals[i])]
        filtered_df = pd.concat([filtered_df, temp_df])

    # Create the data for the API
    outcomes_list = []
    for i in range(len(filtered_df)):
        outcomes_dict = {}
        outcomes_dict["marketExternalId"] = int(filtered_df.iloc[i]['marketExternalId'])
        outcomes_dict["outcomeExternalId"] = int(filtered_df.iloc[i]['outcomeExternalId'])
        outcomes_list.append(outcomes_dict)

    data = {"classExternalId": int(filtered_df['classExternalId'].iloc[0]),
            "competitionExternalId": int(filtered_df['competitionExternalId'].iloc[0]),
            "eventExternalId": int(filtered_df['eventExternalId'].iloc[0]),
            "outcomesExternalIds": outcomes_list}

    return data

# Create function to call the API and return the price
def call_sgm_sportsbet(data, player_names, number_of_disposals):
    
    # Get original price information
    if len(player_names) != len(number_of_disposals):
        raise ValueError('Both lists should have the same length')
    
    filtered_df = pd.DataFrame()
    for i in range(len(player_names)):
        temp_df = data[(data['player_name'] == player_names[i]) & 
                                  (data['number_of_disposals'] == number_of_disposals[i])]
        filtered_df = pd.concat([filtered_df, temp_df])
        
    # Unadjusted price
    unadjusted_price = filtered_df['price'].prod()
    
    data = get_sgm_sportsbet(data, player_names, number_of_disposals)
    url = "https://www.sportsbet.com.au/apigw/multi-pricer/combinations/price"
    headers = {"Content-Type": "application/json",
               'User-Agent': 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Mobile Safari/537.36'}
    response = requests.post(url, data=json.dumps(data), headers=headers)
    adjusted_price = 1 + (response.json()['price']['numerator'] / response.json()['price']['denominator'])
    adjustment_factor = adjusted_price / unadjusted_price
    combined_list = [f"{item1}: {item2}" for item1, item2 in zip(player_names, number_of_disposals)]
    player_string = ", ".join(combined_list)
    
    # Create pandas dataframe of output data
    output_data = {
    'Selections': player_string,
    'Unadjusted Price': unadjusted_price,
    'Adjusted Price': adjusted_price,
    'Adjustment Factor': adjustment_factor,
    'Agency': 'Sportsbet'
    }
    
    return pd.DataFrame(output_data, index=[0])

File: Apps/SGM/sportsbet_sgm.R
library(httr)
library(jsonlite)
library(dplyr)
library(purrr)
library(mongolite)
uri <- Sys.getenv("mongodb_connection_string")

# Sportsbet SGM-----------------------------------------------------------------
sbsgm_con <- mongo(collection = "Sportsbet-SGM", db = "Odds", url = uri)
sportsbet_sgm <- sbsgm_con$find('{}') |> tibble()

sportsbet_sgm <-
  rename(
    sportsbet_sgm,
    eventExternalId = match_id,
    competitionExternalId = comp_id,
    classExternalId = class_id,
    marketExternalId = market_id,
    outcomeExternalId = player_id
  )

#==============================================================================
# Function to get SGM data
#=-=============================================================================

get_sgm_sportsbet <- function(data, player_names, disposal_counts) {
  if (length(player_names) != length(disposal_counts)) {
    stop("Both lists should have the same length")
  }
  
  filtered_df <- data.frame()
  for (i in seq_along(player_names)) {
    temp_df <- data %>%
      filter(player_name == player_names[i],
             number_of_disposals == disposal_counts[i])
    filtered_df <- bind_rows(filtered_df, temp_df)
  }
  
  outcomes_list <- lapply(1:nrow(filtered_df), function(i) {
    list(marketExternalId = as.integer(filtered_df$marketExternalId[i]),
         outcomeExternalId = as.integer(filtered_df$outcomeExternalId[i]))
  })
  
  payload <- list(
    classExternalId = as.integer(filtered_df$classExternalId[1]),
    competitionExternalId = as.integer(filtered_df$competitionExternalId[1]),
    eventExternalId = as.integer(filtered_df$eventExternalId[1]),
    outcomesExternalIds = outcomes_list
  )
  
  return(payload)
}

#==============================================================================
# Make Post Request
#==============================================================================

call_sgm_sportsbet <- function(data, player_names, disposal_counts) {
  if (length(player_names) != length(disposal_counts)) {
    stop("Both lists should have the same length")
  }
  
  filtered_df <- data.frame()
  for (i in seq_along(player_names)) {
    temp_df <- data %>%
      filter(player_name == player_names[i],
             number_of_disposals == disposal_counts[i])
    filtered_df <- bind_rows(filtered_df, temp_df)
  }
  
  unadjusted_price <- prod(filtered_df$price)
  
  payload <- get_sgm_sportsbet(data, player_names, disposal_counts)
  
  url <- 'https://www.sportsbet.com.au/apigw/multi-pricer/combinations/price'
  
  headers <- c('User-Agent' = 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Mobile Safari/537.36',
               'Content-Type' = 'application/json;charset=UTF-8')
  
  response <- POST(url, body = toJSON(payload, auto_unbox = TRUE), add_headers(.headers = headers), encode = "json")
  
  # Check if the request was successful
  if (http_error(response)) {
    stop("API request failed: ", content(response, "text"))
  }
  
  response_content <- content(response, "parsed")
  
  # Check if the response contains the expected data
  if (!"price" %in% names(response_content)) {
    stop("Unexpected API response: 'price' not found")
  }
  
  adjusted_price <- 1 + (response_content$price$numerator / response_content$price$denominator)
  adjustment_factor <- adjusted_price / unadjusted_price
  
  combined_list <- paste(player_names, disposal_counts, sep = ": ")
  player_string <- paste(combined_list, collapse = ", ")
  
  output_data <- data.frame(
    Selections = player_string,
    Unadjusted_Price = unadjusted_price,
    Adjusted_Price = adjusted_price,
    Adjustment_Factor = adjustment_factor,
    Agency = 'Sportsbet'
  )
  
  return(output_data)
}
